command/agent/config_test.go

```diff
@@ -1068,3 +1071,109 @@ func TestReadConfigPaths_dir(t *testing.T) {
 		t.Fatalf("bad: %#v", config)
 	}
 }
+
+func TestUnixSockets(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.SkipNow()
+	}
+
+	usr, err := user.Current()
+	if err != nil {
+		t.Fatal("Could not get current user")
+	}
+
+	tempdir, err := ioutil.TempDir("", "consul-test-")
+	if err != nil {
+		t.Fatal("Could not create a working directory")
+	}
+
+	type SocketTestData struct {
+		Path string
+		Uid  string
+		Gid  string
+		Mode string
+	}
+
+	testUnixSocketPopulation := func(s SocketTestData) (*UnixSocket, error) {
+		return populateUnixSocket("unix://" + s.Path + ";" + s.Uid + ";" + s.Gid + ";" + s.Mode)
+	}
+
+	testUnixSocketPermissions := func(s SocketTestData) error {
+		return adjustUnixSocketPermissions("unix://" + s.Path + ";" + s.Uid + ";" + s.Gid + ";" + s.Mode)
+	}
+
+	_, err = populateUnixSocket("tcp://abc123")
+	if err == nil {
+		t.Fatal("Should have rejected invalid scheme")
+	}
+
+	_, err = populateUnixSocket("unix://x;y;z")
+	if err == nil {
+		t.Fatal("Should have rejected invalid number of parameters in Unix socket definition")
+	}
+
+	std := SocketTestData{
+		Path: tempdir + "/unix-config-test.sock",
+		Uid:  usr.Uid,
+		Gid:  usr.Gid,
+		Mode: "640",
+	}
+
+	std.Uid = "orasdfdsnfoinweroiu"
+	_, err = testUnixSocketPopulation(std)
+	if err == nil {
+		t.Fatal("Did not error on invalid username")
+	}
+
+	std.Uid = usr.Username
+	std.Gid = "foinfphawepofhewof"
+	_, err = testUnixSocketPopulation(std)
+	if err == nil {
+		t.Fatal("Did not error on invalid group (a name, must be gid)")
+	}
+
+	std.Gid = usr.Gid
+	std.Mode = "999"
+	_, err = testUnixSocketPopulation(std)
+	if err == nil {
+		t.Fatal("Did not error on invalid socket mode")
+	}
+
+	std.Uid = usr.Username
+	std.Mode = "640"
+	_, err = testUnixSocketPopulation(std)
+	if err != nil {
+		t.Fatal("Unix socket test failed for no obvious reason (using username)")
```

Comment:
We should definitely expose these errors by including them in the error message.